Direct Consumer stale request filtering and fetcher simplification

joe-redpanda opened the pull request:
This PR does two things.
1. It adds stale fetch filtering to direct_consumer
2. It adds utility methods to diff the epochs used in a fetch request against the epochs found when the request ends
  a. This allows for substantial code simplification.

Direct consumer now tracks its subscriptions with a subscription_epoch. This field functions analogously to assignment_epoch within direct_consumer's fetchers. All fetches will bear the subscription_epoch from the point in time that they were fetched, and on calling fetch_next, the subscription epochs are diff'd against the current subscription epochs to identify and remove stale fetches.

For fetcher cleanup, this PR adds
```cpp
    bool is_consistent_fetcher_epoch(
      const model::topic& topic,
      model::partition_id partition_id,
      const topic_partition_map<epoch_set>& epochs);
```

When a fetch request is started, we snapshot the epochs of all partitions involved in the fetch request. When a fetch response is being handled, the assignment for a given partitions may have been removed, or updated.

This function true iff both the current epoch, and the epoch snapshotted for the request exist and are equal.

This allows for
1. all inconsistent responses to be removed when a fetch_response is handled
2. for a substantially smaller code footprint in the many cases where epoch drift needs to be detected

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Improvements

* Removes stale fetches from direct_consumer output when an ntp has been unassigned and reassigned
* direct_consumer/fetcher code simplification

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -152,6 +169,9 @@ class direct_consumer {
     std::unique_ptr<data_queue> _fetched_data_queue;
     ss::condition_variable _data_available;
 
+    // inc on subscription changes to guard against vending stale data
```
better comment

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -150,7 +151,10 @@ ss::future<fetcher::partitions_with_epoch> fetcher::collect_partitions() {
               partition_fetch_state& fetch_state = p_fs.second;
               ret.assignment_epochs[to_process.topic][fetch_state.partition_id]
-                = fetch_state.assignment_epoch;
+                = epoch_set{
```
remove default constructor

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -303,7 +307,7 @@ bool fetcher::maybe_update_fetch_offset(
     }
 
-    const auto response_epoch = maybe_response_epoch.value();
+    const auto response_epoch = maybe_epochs.value().fetcher_epoch;
```
rename

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -428,10 +432,10 @@ ss::future<> fetcher::do_fetch() {
 }
 
-std::optional<fetcher::assignment_epoch> fetcher::find_assignment_epoch(
+std::optional<fetcher::epoch_set> fetcher::find_assignment_epoch(
```
rename

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -494,6 +498,16 @@ fetcher::process_fetch_response(
             part_data.error = part_response.error_code;
             part_data.partition_id = part_response.partition_index;
 
+            auto partition_epochs
```
split into should_split function

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -630,7 +644,8 @@ fetcher::process_fetch_response(
                     }
 
-                    const auto request_epoch = maybe_request_epoch.value();
+                    const auto request_epoch
```
rename

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -851,7 +868,8 @@ ss::future<> fetcher::assign_partition(
       .assignment_epoch = next_epoch(),
-      .incremental_include = true};
+      .incremental_include = true,
+      .subscription_epoch = subscription_epoch};
```
remove default constructor

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -191,9 +193,11 @@ class fetcher {
         assignment_epoch assignment_epoch{0};
         bool incremental_include{false};
-
+        // differentiate different assignments to the direct consumer as a whole
```
comment improvement

joe-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -31,7 +35,46 @@ direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+                  auto maybe_current_subscription_epoch
+                    = find_subscription_epoch(topic, data.partition_id);
+
+                  this->_cluster->logger().info(
```
we should probably also update the reponse size given that we're shaving data off.

I guess thats a question for the group, how much does having those numbers be correct matter?

joe-redpanda committed:
direct_consumer: api type updates

Adds a new epoch type to track instances of subscription within direct
consumer.

Adds subscription_epoch to fetched_partition_data
Adds size_bytes to index the size of fetched_partition_data's data
member.

joe-redpanda committed:
direct_consumer: add epoch filter to dc

Direct consumer now maintains a subscription_epoch for each partition it
is subscribed to. This epoch is incremented for each partition, and can
be used to differentiate instances of a subscription in the case that a
partition is unsubscribed and resubscribed to the same direct_consumer.

This allows the direct_consumer to filter stale subscriptions from its
returned data set on fetch_next, ensuring the upstream user will not
receive non-monatonic fetches if they unsubscribe and resubscribe a
partition at a lower offset.

This commit adds the subscription epoch and filtering logic to the
direct_consumer, alongside wiring into the plumbing for
subscription_epoch already performed in fetcher.

joe-redpanda committed:
dc/fetcher: update partition_fetch_state

Adds subscription_epoch to partition_fetch_state.

Adds opinionated constructor to make it more difficult to
forget fields. Adds reasonable defaults.

joe-redpanda committed:
dc/fetcher: add epoch_set

Adds epoch_set to contain both subscription_epoch and assignment_epoch.
Will be used in subsequent commits.

joe-redpanda committed:
fetcher: add size to partiton_data

Adds size calculation to fetched_partition_data.

joe-redpanda committed:
direct_consumer: add epoch filtering test

adds testing on stale subscription filtering

joe-redpanda committed:
dc/fetcher: add utility functions

Adding utility functions in preparation for fetcher cleanup.

joe-redpanda committed:
dc/fetcher: filter fetch response

This adds a filter of all tp responses to filter to only responses which
are consistent.

Given that a consistent tp is defined as one where
1. the tp is assigned (in the assigned partitions map)
2. the tp's epochs are present in the epoch map (the tp was assigned at
   the beginning of the request)
3. the fetcher epochs of both sources of truth agree
Any consistent response is one which is not stale, and any insconsistent
response is stale.

All stale responses can be simply ignored.

joe-redpanda committed:
dc/fetcher: simplify inc fetch update logic

process_fetch_response previously checked epoch consistency inline in
the logic which updates incremental_include per the fetch response. This
commit removes the inline check in favor of is_consistent_fetcher_epoch.

joe-redpanda committed:
dc/fetcher: simplify maybe_update_fetch_offset

maybe_update_fetch_offset can skip epoch checks now that all responses
are guaranteed to be consistent at the time at which
maybe_update_fetch_offset is called.

joe-redpanda committed:
dc/fetcher: remove inconsistent code pathway

Fetcher responses are now filtered to only those for consistent
partitions. This removes a defensive 'value_or' on subscription_epoch
assignment, which was needed only in the case of an inconsistent
response.

joe-redpanda committed:
dc/fetcher: simplify do_list_offsets_request

Use is_consistent_fetcher_epoch to simplify do_list_offsets_request

joe-redpanda committed:
dc/fetcher: assign and unassign safety logs

Adds logs for unexpected state during unassign and reassign.

For assign, logs if the ongoing assignment will overwrite a previous
assignment.

For unassign, logs if the ongoing unassign did not find the partition
assignment as expected.

joe-redpanda commented:
Dont review yet, I'm going to drill this with tests to guarantee I didn't introduce any regressions in my cleanup
```
/ci-repeat 15
skip-redpanda-build
skip-units
tests/rptest/direct_consumer_tests/direct_consumer_test.py
dt-repeat=10
```

joe-redpanda commented:
/ci-repeat 15
skip-redpanda-build
skip-units
tests/rptest/direct_consumer_tests/direct_consumer_test.py
dt-repeat=10

joe-redpanda commented:
/ci-repeat 15
skip-redpanda-build
skip-units
tests/rptest/direct_consumer_tests/direct_consumer_test.py
dt-repeat=10

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -887,11 +888,10 @@ ss::future<> fetcher::assign_partition(
+    _partitions[tp.topic].insert_or_assign(
+      tp.partition,
+      partition_fetch_state(
+        tp.partition, offset, next_epoch(), subscription_epoch(-1)));
```
> subscription_epoch(-1))

Its a uint64_t? Also whats the semantic meaning of initializing with a -1, I'd assume this should be plumbed from DC?

bharathv replied:
Oic, this was plumbed in a later commit.

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -152,6 +175,9 @@ class direct_consumer {
     ss::condition_variable _data_available;
 
+    // tracks the version of a subscribed ntp
+    subscription_epoch epoch{0};
```
nit: consider adding a detailed comment on how this epoch plugs into filtering and correctness (especially given so many epochs flying around)

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -505,6 +505,15 @@ fetcher::process_fetch_response(
+              = find_epoch_set(
+                  topic_data.topic, part_response.partition_index, epochs)
+                  .value_or(
+                    epoch_set{fetcher_epoch(-1), subscription_epoch(-1)});
```
there is an overflow with (-1), good to avoid it.

joe-redpanda replied:
intentional to provide an 'invalid' epoch set.

This was a stopgap until a later commit does away with value_or

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -31,12 +37,75 @@ direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+            co_return maybe_response_to_filter;
+        }
+
+        // the remainder is synchronous, no lock is needed
```
curious if the synchronization matters tho? (since the epoch is only monotonically increasing and we only care about equality).

joe-redpanda replied:
I would probably lock it if it were async just to make sure that we're working with a consistent snapshot of _subscriptions for the duration of a filter

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -31,12 +37,75 @@ direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+    for (auto& topic_data : responses_to_filter) {
+        auto& partition_data = topic_data.partitions;
+
+        auto non_stale_subsegment = std::ranges::partition(
```
nit: a simple for loop would be easier to read?

joe-redpanda replied:
chunked_vector doesn't allow single element erase, the best way to erase multiple elements from a chunked vector that I know of is to partition into [keep, discard] and do an erase_to_end

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -31,12 +37,75 @@ direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+              if (is_not_stale) {
+                  decrement_holder.cancel();
+              }
+              return is_not_stale;
```
This seems equivalent to

data.subscription_epoch == maybe_current_subscription_epoch

https://en.cppreference.com/w/cpp/utility/optional/operator_cmp.html

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -270,7 +320,7 @@ class fetcher {
     ss::gate _gate;
     mutex _state_lock;
-    assignment_epoch _epoch{0};
+    fetcher_epoch _epoch{0};
```
nit: consider adding a comment

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -994,6 +938,49 @@ fetcher::do_list_offsets(list_offsets_request req) {
+bool fetcher::is_consistent_fetcher_epoch(
+  const model::topic& topic,
+  model::partition_id partition_id,
+  const topic_partition_map<epoch_set>& epochs) {
+    // not found in epochs -> inconsistent
+    // not found in assignments -> inconsistent
+    // epochs fetcher epoch != assignments fetcher epoch -> inconsistent
+    // epochs fetcher epoch == assignments fetch epoch -> consistent
+
+    auto maybe_epoch_set = find_epoch_set(topic, partition_id, epochs);
+
+    if (!maybe_epoch_set) {
+        return false;
+    }
+    auto epoch_set = *maybe_epoch_set;
+
+    auto maybe_fetch_state = find_fetcher_state(topic, partition_id);
+    if (!maybe_fetch_state) {
+        return false;
+    }
+
+    auto& fetch_state = maybe_fetch_state->get();
+
+    return fetch_state.fetcher_epoch == epoch_set.fetcher_epoch;
+}
```
nit: extraneous new lines.

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -887,15 +813,29 @@ ss::future<> fetcher::assign_partition(
-      .fetch_offset = offset,
-      .assignment_epoch = next_epoch(),
-      .incremental_include = true};
+    auto maybe_existing_assignment = find_fetcher_state(tp.topic, tp.partition);
```
nit: update log line above to add subscription epoch.

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -887,15 +813,29 @@ ss::future<> fetcher::assign_partition(
+    if (maybe_existing_assignment) {
+        auto& existing_assignment = maybe_existing_assignment->get();
+        vlog(
+          logger().warn,
```
nit: wondering if this is really a "warn".. I think its an info at best but may even be a debug IMO.

joe-redpanda replied:
This would require calling assign twice on the same direct consumer for the same ntp. afaik we're resetting offsets with unassign / reassign? in which case this would only get hit unintentionally

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -928,7 +868,11 @@ fetcher::unassign_partition(model::topic_partition_view tp_v) {
-        // partition not found, nothing to unassign
+        vlog(
+          logger().warn,
+          "[broker: {}] Unassign called on tp: {} which is now owned",
```
> now owned

not assigned you mean?

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -797,58 +750,29 @@ ss::future<kafka::error_code> fetcher::maybe_initialise_fetch_offsets(
+            auto& fetch_state = find_fetcher_state(
+                                  response_topic.topic,
+                                  response_partition.partition_id)
+                                  ->get();
```
nit: consider making this a vassert() .. (incase is_consistent_... definition changes in the future).

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -472,6 +435,22 @@ fetcher::process_fetch_response(
+    // we allow for assignment updates to occur while a fetch is ongoing s.t.
+    // assignment updates are not blocked by a longstanding fetch. At this
+    // point, all inconsistent fetch responses should be discarded
+    for (auto& topic_response : resp.data.responses) {
```
I'm wondering why this logic is not folded into the loop below 🤔

joe-redpanda replied:
Could but the separation was intentional.

I thought it was easier to understand if filtering wasn't entwined in the same loop that does processing

bharathv commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_test.cc:
```diff
@@ -468,6 +469,85 @@ TEST_F(consumer_test_mock, TestLeadershipChange) {
 }
 
+/**
+Checks stale subscription filtering.
```
also consider randomzing assign offsets a bit for the test, cases like ABA, A B C (where B < A and C > A) etc.

dotnwat replied:
multi-line comments are prefixed by `*`

bharathv reviewed (commented):
did a first pass, approach lgtm generally.

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -187,12 +187,34 @@ class fetcher {
 private:
     using assignment_epoch = named_type<uint64_t, struct fetcher_epoch_tag>;
     struct partition_fetch_state {
+        partition_fetch_state() = delete;
```
you can drop this the default is not generated because you have an explicit constructor defined

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -187,12 +187,34 @@ class fetcher {
+        partition_fetch_state& operator=(const partition_fetch_state&)
+          = default;
+        partition_fetch_state& operator=(partition_fetch_state&&) = default;
+        ~partition_fetch_state() = default;
```
generally it's fine to drop these if they are all default. you see in the tree a lot of cases where they are all specified, but that is typically because one or more are explicitly deleted and so we specify all of them to make clang-tidy happy.

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -187,12 +187,34 @@ class fetcher {
+          , high_watermark{std::nullopt}
+          , current_leader_epoch{kafka::invalid_leader_epoch}
+          , assignment_epoch{fetcher_epoch}
+          , incremental_include{true}
```
looks like they default was false, now it's true. that deserves a comment. ideally you'd pull it out into separate commit and explain.

bharathv replied:
IIUC, this is not really a concern, right? fetcher.cc always initialized partition_fetch_state{}  incremental_include=true? effectively nothing changed?

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -232,6 +232,27 @@ class fetcher {
     };
+
+    struct epoch_set {
+        epoch_set() = delete;
```
same comment as in previous commit about these

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -150,8 +150,10 @@ ss::future<fetcher::partitions_with_epoch> fetcher::collect_partitions() {
           partitions,
```
```
assignment_epoch name change to fetcher_epoch to better disambiguate

All topic_partition_maps which previously tracked fetcher_epoch now
track both fetcher_epoch and subscription_epoch within an epoch_set

All functions which previously dealt with fetcher_epoch now deal with
epoch_set

Function and variable name changes to better describe their interactions
with epoch_set.
```

Sounds like four sepatate commits

dotnwat commented on src/v/kafka/client/direct_consumer/direct_consumer.h:
```diff
@@ -125,8 +126,24 @@ class direct_consumer {
 
 private:
     struct subscription {
+        subscription() = delete;
```
here too

dotnwat commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -31,12 +37,75 @@ direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+              // decrement a stale partition's contribution to the topic size
+              const auto partition_size = data.size_bytes;
+              auto decrement_holder = ss::defer([&topic_data, partition_size] {
+                  topic_data.total_bytes -= partition_size;
```
> total_bytes

it doesn't look like this value is ever used before the lambda returns, so it does it really need to be deferred until return?

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -271,6 +271,15 @@ class fetcher {
       model::partition_id partition,
       const topic_partition_map<epoch_set>& epochs);
 
+    std::optional<std::reference_wrapper<partition_fetch_state>>
+    find_fetcher_state(
+      const model::topic& topic, model::partition_id partition);
+
+    bool is_consistent_fetcher_epoch(
+      const model::topic& topic,
+      model::partition_id partition_id,
+      const topic_partition_map<epoch_set>& epochs);
+
```
In your commit message

```
Adding utility functions in preparation for fetcher cleanup.

Adds find_fetcher_state which finds a fetcher_state within the map of
assigned partitions, returning a reference to its fetch state if found.

Adds is_consistent_fetcher_epoch, which given a tp and an epoch map,
compares the current fetcher epoch against that from the provided map.
If either epoch is missing, or the epochs disagree, the tp is
inconsistent.
```

The sentence one is fine as the commit message. The later two should be comments in the code on the functions being added. Add any other higher-order context to commit message necessary to understand the context.

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -483,14 +483,14 @@ fetcher::process_fetch_response(
             part_data.error = part_response.error_code;
             part_data.partition_id = part_response.partition_index;
 
```
```
dc/fetcher: clean up subscription_epoch
Cleans up subscription_epoch assignment per response filtering.
```

Please update commit message to state what "clean up" means. Is it a clang-format change, simplified form for the same compuation, etc...

joe-redpanda committed:
dc/fetcher: assignment_epoch->fetcher_epoch

Rename assignment_epoch to fetcher_epoch for clarity. Updates function
names, logs, and comment references as well.

joe-redpanda committed:
dc/fetcher: fetcher_epoch to epoch_set in maps

Changes map values from fetcher_epoch to epoch_set. Variable names
updated accordingly.

joe-redpanda committed:
dc/fetcher: swap function signatures to epoch_set

Updates functions taking and returning fetcher_epoch or
optional<fetcher_epoch> to use epoch_set instead

Variables names updated to new type

joe-redpanda committed:
dc/fetcher: plumb subscription_epoch

Assign subscription_epoch on processing fetch response.

joe-redpanda commented:
in draft until final mop up is complete

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--af69f590-af66-4d00-bf64-e51e155ba8ad-->
<!--TEST RESULTS-->
<!--dd8021db-5853-42b4-bbc9-20fd034ec520-->
<!--TEST RESULTS-->
<!--b6ef7564-936c-47b3-9c57-4f4f59005f33-->
<!--TEST RESULTS-->
<!--f94f0dca-7151-4e40-9720-2b7ad9f05443-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72512">build#72512</a></summary>


| test_class                               | test_method                         | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------------|:------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitPartitionBalancerTests | test_partition_balancer_with_limits | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72512#01995a36-64f0-4c26-8429-ffcedec9c34e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| SchemaEvolutionE2ETests                  | test_legal_schema_evolution         | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "drop_column"}  | integration | https://buildkite.com/redpanda/redpanda/builds/72512#01995a36-64ef-4356-ae00-2d961f640b40 | FLAKY         | 20/21    | upstream reliability is '99.71671388101983'. current run reliability is '95.23809523809523'. drift is 4.47862 and the allowed drift is set to 50. The test should PASS |
| PartitionBalancerTest                    | test_recovery_mode_rebalance_finish | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72512#01995a36-64f3-4433-b984-15bb5592a259 | FLAKY         | 16/21    | upstream reliability is '81.92567567567568'. current run reliability is '76.19047619047619'. drift is 5.7352 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest                 | test_node_operations                | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72512#01995a36-64ea-433d-aced-a5a39398b5a6 | FLAKY         | 19/21    | upstream reliability is '99.19354838709677'. current run reliability is '90.47619047619048'. drift is 8.71736 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72550">build#72550</a></summary>


| test_class                           | test_method                | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72550#01995df7-c21e-4d4f-958b-8e9f81bffe61 | FLAKY         | 17/21    | upstream reliability is '91.19170984455958'. current run reliability is '80.95238095238095'. drift is 10.23933 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72550#01995df9-a302-4bb4-a44c-c91410591c42 | FLAKY         | 20/21    | upstream reliability is '99.53271028037383'. current run reliability is '95.23809523809523'. drift is 4.29462 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                   |
| SimpleEndToEndTest                   | test_consumer_interruption | null                                                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72550#01995df9-a301-4fb5-acad-279376641c54 | FLAKY         | 20/21    | upstream reliability is '98.55072463768117'. current run reliability is '95.23809523809523'. drift is 3.31263 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SimpleEndToEndTest&test_method=test_consumer_interruption                   |
| WriteCachingFailureInjectionTest     | test_unavoidable_data_loss | null                                                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72550#01995df9-a300-4e42-924a-7b885a6a7137 | FLAKY         | 19/21    | upstream reliability is '98.10426540284361'. current run reliability is '90.47619047619048'. drift is 7.62807 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72783">build#72783</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72783#01997887-bc83-4663-85e2-c3a2885e31e0 | FLAKY         | 20/21    | upstream reliability is '99.51845906902086'. current run reliability is '95.23809523809523'. drift is 4.28036 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72876">build#72876</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                       |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72876#01997d27-9929-4d2c-84c6-4849fb9ff447 | FLAKY         | 13/21    | upstream reliability is '82.7433628318584'. current run reliability is '61.904761904761905'. drift is 20.8386 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write |
| DisablingPartitionsTest   | test_disable  | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72876#01997d26-303d-43a4-a7c5-562ab2c39dcf | FLAKY         | 14/21    | upstream reliability is '83.78887841658812'. current run reliability is '66.66666666666666'. drift is 17.12221 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable |
</details>
<!--/body-->

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -483,14 +483,14 @@ fetcher::process_fetch_response(
             part_data.error = part_response.error_code;
```
> dc/fetcher: clean up subscription_epoch assign
>
>Fetcher responses are now filtered to only those for consistent
partitions. This removes a defensive 'value_or' on subscription_epoch
assignment, which was needed only in the case of an inconsistent
response.

It sounds like the subject should be

> dc/fetcher: filter fetcher response by consistent partition

that is, "cleanup" is kind of a content-free description.

joe-redpanda replied:
Filtering was done in the earlier commit, but I see your point. I'll swap this to "remove inconsistent branch"

dotnwat replied:
thanks! the general rule of thumb is that `git log --oneline` should produce a log that contains a high level description of what's happening

dotnwat reviewed (commented):
thanks for the cleanup

bharathv commented:
sorry for the delay here, I'd make another pass today.

bharathv reviewed (approved):
nice work, lgtm.

joe-redpanda commented:
>IIUC, this is not really a concern, right? fetcher.cc always initialized partition_fetch_state{} incremental_include=true? effectively nothing changed?
@bharathv 
The initializer initially set this to 'false', but every invocation of the constructor set it to true.
This change just makes it harder to forget to include the partition in the next fetch
